src/fides/api/util/tcf_util.py

```diff
@@ -0,0 +1,623 @@
+        )
+
+    unique_features: Set[str] = set()
+    for feature_name in record.features:
+        if feature_name in relevant_uses_or_features:
+            unique_features.add(feature_name)
+    return unique_features
+
+
+def _add_top_level_record_to_purpose_or_feature_section(
+    matching_purpose_or_feature_map: Dict[int, TCFPurposeOrFeature],
+    is_purpose_type: bool,
+    use_or_feature: str,
+) -> Optional[TCFPurposeOrFeature]:
+    """
+    Create a purpose or feature record, and add to the top-level [special] purposes or
+    [special] features section if applicable.
+
+    "matching_purpose_or_feature_map" is updated here, in-place.
+    """
+
+    tcf_section_type: Union[Type[TCFPurposeRecord], Type[TCFFeatureRecord]] = (
+        TCFPurposeRecord if is_purpose_type else TCFFeatureRecord
+    )
+    get_gvl_record: Callable = (
+        data_use_to_purpose if is_purpose_type else feature_name_to_feature
+    )
+
+    # Get the matching GVL record for the [special] purpose or [special] feature
+    fideslang_gvl_record: Union[Purpose, Feature] = get_gvl_record(use_or_feature)
+    if not fideslang_gvl_record:
+        return None
+
+    # Transform the base gvl record into the TCF record type that has more elements for TCF display.
+    # Will be a top-level section.
+    top_level_tcf_record: TCFPurposeOrFeature = tcf_section_type(
+        **fideslang_gvl_record.dict()
+    )
+
+    # Add the TCF record to the top-level section in-place if it does not exist
+    if top_level_tcf_record.id not in matching_purpose_or_feature_map:
+        matching_purpose_or_feature_map[top_level_tcf_record.id] = top_level_tcf_record
+
+    # Return the top-level record, whether we just created it, or it previously existed
+    return matching_purpose_or_feature_map[top_level_tcf_record.id]
+
+
+def _embed_purpose_or_feature_under_system(
+    embedded_tcf_record: TCFPurposeOrFeature,
+    system_section: Union[List[TCFPurposeRecord], List[TCFFeatureRecord]],
+    legal_basis_for_processing: Optional[str],
+) -> None:
+    """
+    Embed a second-level TCF purpose/feature under the systems section.
+
+    The systems section is updated in-place.
+    """
+    existing_embedded_purpose_or_feature_record: Optional[TCFPurposeOrFeature] = next(
+        (
+            tcf_sub_record
+            for tcf_sub_record in system_section
+            if tcf_sub_record.id == embedded_tcf_record.id
+        ),
+        None,
+    )
+
+    if existing_embedded_purpose_or_feature_record:
+        # Update legal_bases on existing embedded purpose or feature record beneath system if applicable
+        _extend_legal_bases(
+            existing_embedded_purpose_or_feature_record,
+            legal_basis_for_processing,
+            replace=False,
+        )
+        return
+    # Nest new cloned TCF purpose or feature record beneath system otherwise
+    system_section.append(embedded_tcf_record)  # type: ignore[arg-type]
```

Comment:
potential update: 
```suggestion
def _embed_purpose_or_feature_under_system(
    embedded_tcf_record: TCFPurposeOrFeature,
    system_section: Union[List[TCFPurposeRecord], List[TCFFeatureRecord]],
    legal_basis_for_processing: Optional[str],
) -> None:
    """
    Embed a second-level TCF purpose/feature under the systems section.

    The systems section is updated in-place.
    """
    existing_embedded_purpose_or_feature_record: Optional[TCFPurposeOrFeature] = next(
        (
            tcf_sub_record
            for tcf_sub_record in system_section
            if tcf_sub_record.id == embedded_tcf_record.id
        ),
        None,
    )

    if existing_embedded_purpose_or_feature_record:
        # Update legal_bases on existing embedded purpose or feature record beneath system if applicable
        _extend_legal_bases(
            existing_embedded_purpose_or_feature_record,
            legal_basis_for_processing,
            replace=False,
        )
        return
    # Nest new cloned TCF purpose or feature record beneath system otherwise
    # Ensuring it has the appropriate legal basis
    _extend_legal_bases(
                embedded_tcf_record, legal_basis_for_processing, replace=True
      )
    system_section.append(embedded_tcf_record)  # type: ignore[arg-type]
```